pubsub/natspubsub/nats.go

```diff
@@ -0,0 +1,292 @@
+func (t *topic) As(i interface{}) bool {
+	c, ok := i.(**nats.Conn)
+	if !ok {
+		return false
+	}
+	*c = t.nc
+	return true
+}
+
+// ErrorAs implements driver.Topic.ErrorAs
+func (*topic) ErrorAs(error, interface{}) bool {
+	return false
+}
+
+// ErrorCode implements driver.Topic.ErrorCode
+func (*topic) ErrorCode(err error) gcerrors.ErrorCode {
+	switch err {
+	case nil:
+		return gcerrors.OK
+	case context.Canceled:
+		return gcerrors.Canceled
+	case errNotInitialized, nats.ErrBadSubject:
+		return gcerrors.FailedPrecondition
+	case nats.ErrAuthorization:
+		return gcerrors.PermissionDenied
+	case nats.ErrMaxPayload, nats.ErrReconnectBufExceeded:
+		return gcerrors.ResourceExhausted
+	}
+	return gcerrors.Unknown
+}
+
+type subscription struct {
+	nc   *nats.Conn
+	nsub *nats.Subscription
+	err  error
+}
+
+// CreateSubscription returns a *pubsub.Subscription representing a NATS subscription.
+// TODO(dlc) - Options for queue groups?
+func CreateSubscription(nc *nats.Conn, subscriptionName string) *pubsub.Subscription {
+	return pubsub.NewSubscription(createSubscription(nc, subscriptionName), nil)
+}
+
+func createSubscription(nc *nats.Conn, subscriptionName string) driver.Subscription {
+	sub, err := nc.SubscribeSync(subscriptionName)
+	return &subscription{nc, sub, err}
+}
+
+// ReceiveBatch implements driver.ReceiveBatch.
+func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
+	if s == nil || s.nsub == nil {
+		return nil, nats.ErrBadSubscription
+	}
+
+	// Return right away if the ctx has an error.
+	if err := ctx.Err(); err != nil {
+		return nil, err
+	}
+
+	var ms []*driver.Message
+
+	// We will assume the desired goal is at least one message since the public API only has Receive().
+	// We will load all the messages that are already queued up first.
+	for {
+		msg, err := s.nsub.NextMsg(0)
+		if err == nil {
+			dm, err := decode(msg)
+			if err != nil {
+				return nil, err
+			}
+			ms = append(ms, dm)
+			if len(ms) >= maxMessages {
+				break
+			}
+		} else if err == nats.ErrTimeout {
+			break
+		} else {
+			return nil, err
+		}
+	}
+
+	// If we have anything go ahead and return here.
+	if len(ms) > 0 {
+		return ms, nil
+	}
+
+	// Here we will assume the ctx has a deadline and will allow it to do its thing. We may
+	// get a burst of messages but we will only wait once here and let the next call grab them.
+	// The reason is so that if deadline is not properly set we will not needlessly wait here
+	// for more messages when the user most likely only wants one.
+	msg, err := s.nsub.NextMsgWithContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+	dm, err := decode(msg)
+	if err != nil {
+		return nil, err
+	}
+	ms = append(ms, dm)
+	return ms, nil
+}
+
+// Convert NATS msgs to *driver.Message.
+func decode(msg *nats.Msg) (*driver.Message, error) {
+	if msg == nil {
+		return nil, nats.ErrInvalidMsg
+	}
+	var dm driver.Message
+	// Everything is in the msg.Data
```

Comment:
Similarly here, what if the message was sent from a native NATS client? We'd want it to decode to a `driver.Message` with the `Body` set and no `Metadata`?

Consider adding tests for `encode` and `decode` that verify that encoding a `driver.Message` with empty `Metadata` is just `Message.Body`, and that decoding an arbitrary `[]byte` (that doesn't decode to a `driver.Message`) decodes to a `driver.Message` with the `Body` set to the bytes.

Reply:
I can add a test for this.